### PR TITLE
Don't allow empty memberlist on cluster id changes [API-1215]

### PR DESCRIPTION
**Pull request by sancar**
When cluster id changed we were resetting member list
with an empty memberlist. It turned out that there are implementations
that assumes it will never be empty.

Following test failure is an example of it.
https://github.com/hazelcast/hazelcast/issues/20264

Instead of clearing the memberlist on cluster id change, we
are keeping the cluster id as part of the latest snapshot
to be able to fire correct events. In other words, on cluster restart
we need to fire removed events for all old members and added
events for all new members coming from restarted cluster. 
We were achieving this via clearing memberlist and firing removed
events first. Then after some time new memberlist comes and we fire
added events. And the memberlist is empty during this period. 
To prevent that, we have changed the logic. Now `clusterUuid` is 
part of the logic. We don't clear the memberlist at all. Upon receiving, 
an event, we compare existing and new clusterUuid. If they are different, 
without comparing the members, we fire removed events for all existing 
members, and added events for all new ones to achieve same behaviour. 

Note that clearing memberlist on cluster id change was
introduced as a fix to https://github.com/hazelcast/hazelcast/pull/18245
So it is important not to break related ClientHotRestartTest

fixes https://github.com/hazelcast/hazelcast/issues/20264

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Label `Add to Release Notes` or `Not Release Notes content` set
- [x] Request reviewers if possible
- [ ] Send backports/forwardports if fix needs to be applied to past/future releases
- [x] New public APIs have `@Nonnull/@Nullable` annotations
- [x] New public APIs have `@since` tags in Javadoc

### CODE REVIEW

File: hazelcast/src/test/java/com/hazelcast/client/impl/spi/impl/ClientClusterServiceImplTest.java [hunk 1]
```diff
@@ -0,0 +1,487 @@
...
+            @Override
+            public void memberRemoved(MembershipEvent membershipEvent) {
+                removedCount.incrementAndGet();
+            }
+        });
+
+        //called on cluster restart
+        clusterService.onClusterRestart();
+
+        clusterService.handleMembersViewEvent(1, asList(member("127.0.0.2")), UUID.randomUUID());
```
Comment (utku-caglayan) [inline hazelcast/src/test/java/com/hazelcast/client/impl/spi/impl/ClientClusterServiceImplTest.java:175]: Should we also assert for added members address or UUID to check if the correct member is published in member event?

**Review by utku-caglayan — APPROVED** [review]
Looks great, noted minor suggestions

File: hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java [hunk 2]
```diff
@@ -727,29 +726,26 @@ protected TcpClientConnection createSocketConnection(Address target) {
...
-        return translate(member.getAddress());
+        return translate(member, AddressProvider::translate);
+    }
+
+    private Address translate(Address address) {
+        return translate(address, AddressProvider::translate);
     }
 
-    private Address translate(Address target) {
+    private <T> Address translate(T target, BiFunctionEx<AddressProvider, T, Address> translateFunction) {
```
[the remark this answers is not in the document]
Comment (sancar) [reply hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java:736]: After realizing that we are using the public address of the member only if the `DefaultAddressProvider` is used, removing `TranslateToPublicAddressProvider`(`client.getClientClusterService().translateToPublicAddress())`) and getting the public address from the main logic seemed much clear to me. This is something specific to `DefaultAddressProvider` . 

We also have `RemoteAddressProvider` which will do the translation of address and member differently and has nothing to do with `client.getClientClusterService().translateToPublicAddress())`

File: hazelcast/src/test/java/com/hazelcast/client/impl/spi/impl/ClientClusterServiceImplTest.java (hunk 1 above)
Comment (sancar) [reply hazelcast/src/test/java/com/hazelcast/client/impl/spi/impl/ClientClusterServiceImplTest.java]: Added member checks here https://github.com/hazelcast/hazelcast/pull/20818/commits/7606fc83d49b6bd2f4f7e2b88e401a63e9d46f29

File: hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (utku-caglayan) [reply hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java:736]: Thanks for the explanation!

File: hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/ClientClusterServiceImpl.java [hunk 3]
```diff
@@ -288,35 +235,38 @@ private MemberListSnapshot createSnapshot(int memberListVersion, Collection<Memb
...
         }
 
         List<MembershipEvent> events = new LinkedList<>();
 
         for (Member member : deadMembers) {
-            events.add(new MembershipEvent(client.getCluster(), member, MembershipEvent.MEMBER_REMOVED, currentMembers));
-            Connection connection = connectionManager.getConnection(member.getUuid());
-            if (connection != null) {
-                connection.close(null,
-                        new TargetDisconnectedException("The client has closed the connection to this member,"
```
Comment (ihsandemir) [inline hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/ClientClusterServiceImpl.java:314]: Why did we remove connection close logic here?

File: hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/ClientClusterServiceImpl.java [hunk 4]
```diff
@@ -288,35 +235,38 @@ private MemberListSnapshot createSnapshot(int memberListVersion, Collection<Memb
...
+
+        if (clusterUuid.equals(memberListSnapshot.get().clusterUuid)) {
+            for (Member member : currentMembers) {
+                if (!deadMembers.remove(member)) {
+                    newMembers.add(member);
+                }
             }
+        } else {
+            // if cluster uuid is not same, then we will not try to match the current members to previous members
+            // As a result all previous members are dead and all current members are new members
```
Comment (ihsandemir) [inline hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/ClientClusterServiceImpl.java:259]: Is there a behaviour change here? Will the user receive member removed event for the member in the old cluster? If so, how?

**Review by ihsandemir — COMMENTED** [review]
why did you delete the test `ClientClusterServiceMemberListTest`?

File: hazelcast/src/main/java/com/hazelcast/client/impl/clientside/HazelcastClientInstanceImpl.java [hunk 5]
```diff
@@ -864,7 +870,7 @@ public void onClusterChange() {
 
         dispose(onClusterChangeDisposables);
         //clear the member lists
```
Comment (ihsandemir) [inline hazelcast/src/main/java/com/hazelcast/client/impl/clientside/HazelcastClientInstanceImpl.java:872]: incorrect comment?

File: hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/ClientClusterServiceImpl.java [hunk 6]
```diff
@@ -208,69 +174,50 @@ public void waitInitialMemberListFetched() {
         }
     }
 
-    public void clearMemberListVersion() {
+    public void onClusterRestart() {
```
Comment (ihsandemir) [inline hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/ClientClusterServiceImpl.java:177]: I think the names `onClusterRestart` and `onClusterChange` 's confusing. Same methods also exist at HazelcastClientInstanceImpl.

File: hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/ClientClusterServiceImpl.java [hunk 7]
```diff
@@ -208,69 +174,50 @@ public void waitInitialMemberListFetched() {
...
             MemberListSnapshot clusterViewSnapshot = memberListSnapshot.get();
-            // This check is necessary so that when handling auth response, it will not
-            // intervene with client failover logic
-            if (clusterViewSnapshot != EMPTY_SNAPSHOT) {
-                memberListSnapshot.set(new MemberListSnapshot(0, clusterViewSnapshot.members));
+            // This check is necessary so in order not to override changing cluster information when:
+            // - registering cluster view listener back to the new cluster.
+            // - on authentication response when cluster uuid change is detected.
+            if (clusterViewSnapshot.version != INITIAL_LIST_VERSION) {
+                memberListSnapshot.set(new MemberListSnapshot(0,
```
Comment (ihsandemir) [inline hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/ClientClusterServiceImpl.java:187]:
```suggestion
                memberListSnapshot.set(new MemberListSnapshot(clusterViewSnapshot.version,
```

File: hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/ClientClusterServiceImpl.java (hunk 4 above)
Comment (sancar) [reply hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/ClientClusterServiceImpl.java:259]: The behaviour change is as descriped in the pr description.
 ```
Now clusterUuid is part of the logic. We don't clear the memberlist at all. Upon receiving,
an event, we compare existing and new clusterUuid. If they are different,
without comparing the members, we fire removed events for all existing
members, and added events for all new ones to achieve same behaviour.
```

**Review by ihsandemir — APPROVED** [review]

**Pull request merged**